File: backend/routers/blog.py
from fastapi import APIRouter, Depends
from backend.models.user_model import UserProfile
from backend.models.blog_model import (
    BlogCreateRequest, BlogUpdateRequest, BlogResponse,
    BlogSummary, CommentResponse, CommentCreate
)
from backend.services.blog_service import BlogService
from backend.middleware.auth_middleware import get_current_user

router = APIRouter(prefix="/blog", tags=["blog"])
blog_service = BlogService()


@router.get("/", response_model=list[BlogSummary])
async def list_blogs(current_user: UserProfile = Depends(get_current_user)):
    return blog_service.get_blogs(current_user.id)


@router.post("/", response_model=BlogResponse)
async def create_blog(body: BlogCreateRequest, current_user: UserProfile = Depends(get_current_user)):
    return blog_service.create_blog(
        current_user.id, body.title, body.content,
        body.cover_image, body.tags, body.published
    )


@router.get("/{blog_id}", response_model=BlogResponse)
async def get_blog(blog_id: str, current_user: UserProfile = Depends(get_current_user)):
    return blog_service.get_blog(blog_id, current_user.id)


@router.put("/{blog_id}", response_model=BlogResponse)
async def update_blog(blog_id: str, body: BlogUpdateRequest, current_user: UserProfile = Depends(get_current_user)):
    return blog_service.update_blog(blog_id, current_user.id, body)


@router.delete("/{blog_id}")
async def delete_blog(blog_id: str, current_user: UserProfile = Depends(get_current_user)):
    return blog_service.delete_blog(blog_id, current_user.id)


@router.post("/{blog_id}/like")
async def toggle_like(blog_id: str, current_user: UserProfile = Depends(get_current_user)):
    return blog_service.toggle_like(blog_id, current_user.id)


@router.get("/{blog_id}/comments", response_model=list[CommentResponse])
async def get_comments(blog_id: str, current_user: UserProfile = Depends(get_current_user)):
    return blog_service.get_comments(blog_id)


@router.post("/{blog_id}/comments", response_model=CommentResponse)
async def add_comment(blog_id: str, body: CommentCreate, current_user: UserProfile = Depends(get_current_user)):
    return blog_service.add_comment(blog_id, current_user.id, body.content)


@router.delete("/comments/{comment_id}")
async def delete_comment(comment_id: str, current_user: UserProfile = Depends(get_current_user)):
    return blog_service.delete_comment(comment_id, current_user.id)

File: frontend/lib/blog.ts
import api from "./axios"

export interface BlogSummary {
  id: string
  title: string
  cover_image?: string
  tags: string[]
  published: boolean
  author_name: string
  likes_count: number
  comments_count: number
  is_liked: boolean
  created_at?: string
}

export interface CommentResponse {
  id: string
  blog_id: string
  user_id: string
  content: string
  author_name: string
  created_at?: string
}

export interface BlogResponse {
  id: string
  title: string
  content: string
  cover_image?: string
  tags: string[]
  published: boolean
  author_id: string
  author_name: string
  likes_count: number
  comments_count: number
  is_liked: boolean
  created_at?: string
  updated_at?: string
  comments: CommentResponse[]
}

export async function getBlogs(): Promise<BlogSummary[]> {
  const res = await api.get<BlogSummary[]>("/blog/")
  return res.data
}

export async function getBlog(id: string): Promise<BlogResponse> {
  const res = await api.get<BlogResponse>(`/blog/${id}`)
  return res.data
}

export async function createBlog(data: {
  title: string
  content: string
  cover_image?: string
  tags?: string[]
  published?: boolean
}): Promise<BlogResponse> {
  const res = await api.post<BlogResponse>("/blog/", data)
  return res.data
}

export async function updateBlog(
  id: string,
  data: {
    title?: string
    content?: string
    cover_image?: string
    tags?: string[]
    published?: boolean
  }
): Promise<BlogResponse> {
  const res = await api.put<BlogResponse>(`/blog/${id}`, data)
  return res.data
}

export async function deleteBlog(id: string): Promise<void> {
  await api.delete(`/blog/${id}`)
}

export async function toggleLike(id: string): Promise<{ liked: boolean; likes_count: number }> {
  const res = await api.post<{ liked: boolean; likes_count: number }>(`/blog/${id}/like`)
  return res.data
}

export async function getComments(id: string): Promise<CommentResponse[]> {
  const res = await api.get<CommentResponse[]>(`/blog/${id}/comments`)
  return res.data
}

export async function addComment(id: string, content: string): Promise<CommentResponse> {
  const res = await api.post<CommentResponse>(`/blog/${id}/comments`, { content })
  return res.data
}

export async function deleteComment(id: string): Promise<void> {
  await api.delete(`/blog/comments/${id}`)
}
